### DISCUSSION TITLE
Support `__eq__` and `__repr__` in Select, Update and others

### DISCUSSION THREAD

**Discussion by rafaelclp** [sqlalchemy/sqlalchemy#11792] [Ideas]
## Context

Testing with SQLAlchemy isn't great.

For simple and straightforward queries, maybe there isn't much to be gained from writing unit tests (and we often don't). But we do have some complex cases, where unit tests are really important.

Here is a very simplified example of how I test these things:
```python
def test_build_query_xyz(self):
    data = ...
    tag = ...
    external_id = ...

    expected_query = (
        select(MyEntity)
        .where(
            and_(
                MyEntity.tag == tag,
                MyEntity.entity_id == build_find_entity_id_from_external_id_and_tag_stmt(external_id, tag).scalar_subquery(),
            )
        )
        .order_by(MyEntity.updated_at)
        .limit(1)
    )

    produced_query = MyEntity.build_query_xyz(data)

    self.assertEqual(str(expected_query), str(produced_query))
```

> **NOTE:** I know `str` is not the best way to do this. I should compile it using the target dialect. But this is good enough for these tests.

This test isn't too hard to read, I actually like it. I have some very complex `build_query_xyz`, which dynamically builds a query from `data`, and I can just assert it builds the correct query from the `data` I passed in.

But let's continue.

## Problem 1 - Writing tests for real code

However, what if I don't want to make an intermediary method called `build_query_xyz` for every query? Who, in fact, builds these intermediary methods? I usually just have `get_xyz`.

```python
async def test_get_xyz(self):
    ...  # prepare data and everything else here

    produced_xyz = await MyEntity.get_xyz(data, session=self.session_mock)

    self.session_mock.execute.assert_awaited_once()
    produced_query = self.session_mock.execute.call_args[0][0]
    self.assertEqual(str(expected_query), str(produced_query))

    ...  # more asserts here
```

Now things are starting to look ugly and hard to read. Tests should be easy to understand. This would be much better:

```python
async def test_get_xyz(self):
    ...  # prepare data and everything else here

    produced_xyz = await MyEntity.get_xyz(data, session=self.session_mock)

    self.session_mock.execute.assert_awaited_once_with(expected_query)

    ...  # more asserts here
```

But of course, this doesn't work, because `Select`, `Update` and others don't implement `__eq__`.

## Problem 2 - Debugging test failures

When the test fails, figuring out why it failed is really hard. In my case, the queries are long and complex.

I can `print(query)`, but figuring out exactly where things went wrong looking at the generated SQL then having to translate the sql back to sqlalchemy makes me wonder why I use sqlalchemy in the first.  This is what `print(query)` would show me for the simple query above:
```
SELECT my_table.entity_id, my_table.tag, my_table.external_id, my_table.version, ...........
FROM my_table 
WHERE my_table.tag = :tag_1 AND my_table.entity_id = (SELECT my_table.entity_id 
FROM my_table 
WHERE my_table.tag = :tag_2 AND my_table.external_id = :external_id_1 AND my_table.version = :version_1 ORDER BY my_table.updated_at
 LIMIT :param_1) ORDER BY my_table.updated_at DESC, my_table.version DESC
```
I want to reiterate that the queries I test are generally not this simple. I was just too lazy to rewrite the table name, class name, and column names for a more complex case to use in this discussion. I chose the simplest I had available, then simplified it even further. I regret this decision now, but I won't go back and rewrite all the examples, so let's pretend this is complex enough to be hard to debug.

It would be so much easier if I could just print a repr of the statement. Currently, with `print(repr(query))`, all I get is `<sqlalchemy.sql.selectable.Select object at 0x105a88fa0>`.

It would be nice to see something like `select(MyEntity).where(...).order_by(MyEntity.updated_at).limit(1)`. I understand that maybe we wouldn't see exactly that (especially `MyEntity` and `MyEntity.updated_at`), but that's okay, as long as it's close enough to how I'd write the actual code.

**Comment by CaselIT** [comment]
Hi,

While potentially supporting `__eq__` could be something to be considered, it's not planned at this time.
You can work around it by providing your own wrapper that does it:
```py
class eq_on_str:
    def __init__(self, expected):
        self.expected = expected
    def __eq__(self, other):
        return str(self.expected) == str(other)
```
You can then use it on `assert_awaited_once_with(eq_on_str(expected_query))`

Also regarding `repr` it's not planned at this time. Doing it in a meaningful way would mean writing another compiler the generates the representation of the strings.
Since all sqlalchemy object support the visitor pattern it would be possible to write an external implementation that does it. Maybe there is already a library that does it too?

**Reply by rafaelclp** [reply]
That's a great idea, thanks.

I still think that at some point it would be very nice to have both by default, as it would make testing good by default. But I understand it's not too hard to build our own simple implementation that works for most cases. Thanks.

**Reply by zzzeek** [reply]
repr doesnt work for a lot of reasons such as this one, unless I'm missing some obvious way to handle this kind of ambiguity with repr():

```py

my_alias = aliased(MyClass)
my_other_alias = aliased(MyClass)


stmt1 = select(my_alias).where(my_alias.c.id == my_other_alias.c.other_id)
stmt2 = select(my_alias).where(my_alias.c.id == my_alias.c.other_id)

# does this assert pass?   how?
assert repr(stmt1) != repr(stmt2)

```

**Comment by zzzeek** [comment]
`__eq__()` itself as the actual method is a non-starter.  SQLAlchemy special purposes this method as a means to generate new SQL expression constructs.   So for those cases where "some SQL = some other SQL" doesn't make sense, it's essential that `__eq__()` continue to raise as it does.      for cases where `=` does evaluate to some new SQL, it's also essential that in most cases that resulting object does not evaluate in a `__bool__` context and continues to raise, since there's more than one way to define this comparison (identity or deep traverse, the latter of which is expensive).

There is now a deep traverse / rich comparison method for SQL constructs which you should use here, which is [compare()](https://docs.sqlalchemy.org/en/20/core/foundation.html#sqlalchemy.sql.expression.ClauseElement.compare).   This method isn't correctly documented right now as it doesn't describe what the means of comparison is, as it has changed in more recent versions of SQLAlchemy starting with 1.4.   however it now does a mutual traversal of both structures, using rich comparison of each element to produce a result, and it builds upon the caching capabilities that were introduced in SQLAlchemy 1.4.  

as for `__repr__()`  generating real code, this would be extremely difficult to implement and would not generate code that's actually runnable.   since your real [XY problem](https://xyproblem.info/) is asserting common structures, use the `.compare()` method.

**Reply by rafaelclp** [reply]
I tried `a.compare(b)`, and it replaces `str(a) == str(b)` pretty well, but doesn't solve either of the underlying problems.

The first problem is about readability. I can already compare queries using `str(a) == str(b)`, but the test code is hard to read. I understand your concern about `__eq__`, it's indeed not viable. With @CaselIT 's idea I can make the test much easier to read, and I'm satisfied with his solution, so I consider the first problem solved.

The second problem is about debugging a broken test. It's great when tests don't break, but when they do, you really do want good feedback on what exactly broke. `compare` only returns `True`/`False`, it doesn't help me at all. Perhaps `__repr__` isn't necessarily the solution, and I understand your concern about `__repr__` and `aliased` in particular. I don't think it has to produce runnable code though - it's great when it does, but has never been a requirement. For aliases, one ugly option is:
- `repr(aliased(MyClass))` -> `'aliased(MyClass)'`
- `repr(select(my_alias).where(my_alias.c.id == my_other_alias.c.other_id))` -> `select(aliased<'alias1'>(MyClass)).where(aliased<'alias1'>(MyClass).id == aliased<'alias2'>(MyClass).other_id)`.

It's not runnable, but it differentiates the aliases at least. Perhaps instead of `__repr__` we could have a `dump()`/`debug_dump()` method or utility function. But I understand the solution above for aliases is far from perfect, and it's definitely not the only case where it's hard or impossible to repr properly. So maybe for now, looking for an external implementation / third-party lib as @CaselIT proposed would be a better bet for me.

All I need is a way to easily see where the queries differ. Maybe instead of `__repr__`, it would indeed be better to traverse the objects. I'd still need a way to display how they differ, and I'm not sure how to do that without something similar to `__repr__`.
